book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+
+In the future, PRQL may define a common physical layout of types, probably using
+Apache Arrow.
+
+## Syntax
+
+```
+# built-in types
+type int
+type float
+type bool
+type text
+type char
+type null
+
+# users-defined types
+type my_int = int
+
+# sum type (union or enum)
+type number = int | float
+type scalar = number | bool | str | char
+
+# by default types are not nullable
+type my_nullable_int = int | null
+
+# user-defined enum
+type open
+type pending
+type closed
+type status = open | pending | closed
+
+# product type
+type my_struct = [id = int_nullable, name = str]
+type my_timestamp_with_timezone = [int, text]
+
+# list type (I'm not sure about this syntax)
+type list = {T}
+
+# relations are list of structs
+type my_relation = {[
+	id = int,
+	title = text,
+	age = int
+]}
+
+type invoices = {[
+    invoice_id = int64,
+    issued_at = timestamp,
+    labels = {text}
+
+    #[repr(json)]
+    items = [{
+        article_id = int64,
+        count = int16 where x -> x >= 1,
+    }],
+    paid_by_user_id = int64 | null,
+    status = status,
+]}
+```
+
+## Built-in primitives
+
+I've mentioned having `int32` and `int64` as distinct types, but initial
+implementation I don't see a need to do that. Instead we can start with a
+built-in `int`, which can later be replaced by `type int = int32 | int64`.
+
+The general rule for "when to make a distinction between types" would be "as
+soon as we find an operation that would be expressed differently". In this
+example, that would require some operation on `int32` to have different syntax
+than same operation over `int64`.
+
+We can have such relaxed rule because PRQL is not aiming to be a Data Definition
+Language and does not have to bother with exact bit layout of types.
+
+## Appendix
+
+### Non-bijection cases between PTS and STS
+
+There are cases where a PTS construct has multiple possible and valid
+representations in some STSs.
+
+For such cases, we'd want to have something similar to Rust's `#[repr(X)]` which
+says "data in this type is represented as X" (we'd probably want a different
+syntax).
+
+This is needed because translation from PRQL operation to SQL may depend on the
+representation.
+
+Using SQLite as an example again, users may have some data stored as INTEGER and
+some as TEXT, but would want to define both of them as PTS `timestamp`. They
+would attach `#[repr(INTEGER)]` or `#[repr(TEXT)]` to the type. This would
+effect how `timestamp - timestamp` is translated into SQL. INTEGER can use
+normal int subtraction, but TEXT must apply `unixepoch` first.
+
+A similar example is a "string array type" in PTS that could be represented by
+an `text[]` (if DBMS supports arrays) or `json` or it's variant `jsonb` in
+Postgres. Again, the representation would affect operators: in Postgres arrays
+can be access with `my_array[1]` and json uses `my_json_array -> 1`. This
+example may not be applicable, if we decide that we want a separate JSON type in
+PST.
```

Comment:
> @vanillajonathan 
> Vendor-specific data types?
> 
> Some databases have non-standard vendor-specific data types, such as jsonbin PostgreSQL. One thing that bothers me about this is that just by looking at a database schema it is difficult to discern whether a data type is a standard data type or a non-standard data type, and to which DBMS it belongs to.
>
> CSS deals with this with vendor prefixes, example Chrome might have their vendor-specific CSS property called -webkit-user-drag and Firefox might have a vendor-specific CSS property named -moz-radial-gradient, then these are denoted as such by the -webkit- and -moz- prefix respectively.
>
> Media types addresses this with "vendor trees", example application/vnd.microsoft.portable-executable.

What I meant here, is that we may be able to support vendor-specific data types jsonb in Postgres by defining them as normal structs and lists in PRQL and mark them as having a "Postgres JSON" representation.

This would tell the compiler to use postgres's operators (i.e. `->`) for implementing standard PRQL operations (in this case field indirection - accessing a field of a struct).